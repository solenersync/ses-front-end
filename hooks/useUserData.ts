import { useEffect, useState } from 'react';
import { useSession } from 'next-auth/react';
import Router from 'next/router';
import { getUser } from 'api/userApi';

export const useUserData = () => {
  const { status, data: sessionData } = useSession();
  const [userId, setUserId] = useState<string | null>(null);

  useEffect(() => {
    if (status === 'loading') {
      return;
    }
    if (!sessionData) {
      Router.replace('/login');
      return;
    }
    const { user } = sessionData;
    async function fetchData() {
      const userData = await getUser(user.email);
      setUserId(userData.userId);
    }
    fetchData();
  }, [status, sessionData]);

  return userId;
};