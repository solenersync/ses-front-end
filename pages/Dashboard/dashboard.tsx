import { AppLayout } from "components/Layouts/AppLayout";
import { NextPageWithLayout } from "types";
import React, { useEffect, useState } from "react";
import SolarForecastChart from "components/SolarForecastChart/solarForecastChart";
import { useUserData } from "hooks/useUserData";

const Dashboard: NextPageWithLayout = () => {
  const [month, setMonth] = useState<number | null>(null);
  const userId = useUserData();

  useEffect(() => {
    setMonth(new Date().getMonth() + 1);
  }, []);

  return (
    <>
      {userId && month ? (
        <SolarForecastChart userId={userId} month={month}></SolarForecastChart>
      ) : null}
    </>
  );
};

Dashboard.getLayout = (page) => {
  return <AppLayout pageTitle="Dashboard">{page}</AppLayout>;
};

export default Dashboard;
