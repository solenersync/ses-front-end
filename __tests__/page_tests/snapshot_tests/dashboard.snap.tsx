import { User} from 'next-auth';
import { useSession as originalUseSession } from 'next-auth/react';
import { render } from '@testing-library/react';
import Dashboard from 'pages/dashboard';
import { expect } from '@jest/globals';
import { useUserData } from 'hooks/useUserData';

jest.mock('hooks/useUserData');
jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));
jest.mock('next-auth/react', () => ({
  ...jest.requireActual('next-auth/react'),
  useSession: jest.fn()
}));
jest.mock('components/SolarForecastChart/solarForecastChart', () => {
  return function MockSolarForecastChart() {
    return <div data-testid='solar-forecast-chart'></div>;
  };
});

describe('Dashboard', () => {
  const user: User = { name: 'John Doe', email: 'jd@test.com',  userId: 1, id:'' };

  beforeEach(() => {
    (useUserData as jest.Mock).mockReturnValue(user);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should render a chart on the dashboard ', async () => {
    const user: User = { name: 'John Doe', email: 'jd@test.com',  userId: 1, id:'' };
    (originalUseSession as jest.Mock).mockReturnValue({ status: 'authenticated', data: user });
    
    const { container } = render(<Dashboard />);
    expect(container).toMatchSnapshot();
  });
});
