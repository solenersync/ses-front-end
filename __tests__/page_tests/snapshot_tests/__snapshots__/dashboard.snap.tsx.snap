// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render a chart on the dashboard  1`] = `
<div>
  <div
    data-testid="solar-forecast-chart"
  />
</div>
`;
